### PR TITLE
TINKERPOP-2076 Build on Java 11

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-2076

Not sure what else to say - finally done. 

All tests pass with `docker/build.sh -t -n -i`

VOTE +1

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,9 +2,9 @@ language: java
 os:
   - linux
 jdk:
-  - openjdk8
+  - openjdk11
```
Comment (jorgebay) [inline .travis.yml:5]: I think we should add a job below using `openjdk8`, that way we know that we are not breaking java 8 support.

**Comment by spmallette** [comment]
> Are we dropping support for jdk 8 altogether?

I think we are dropping support for buliding on jdk8, at least insofar as our maven configuration. You need jdk11 to build going forward. However, we support running on jdk8. 

I'm glad you brought this up because I still had an old configuration in there from some early testing I was doing that had the release version at 11. It needed to be set at 8. 

https://github.com/apache/tinkerpop/pull/1299/commits/23fe7dea89f5d65f6e11b828e5fa5d4d62a68767

The cross-compilation option is described here:

http://openjdk.java.net/jeps/247

> I think we should add a job below using openjdk8, that way we know that we are not breaking java 8 support.

My understanding here is that this setting would catch those situations where we used something incompatible with jdk8 at compile time. In this way we would not need to explicitly build on jdk8. I suppose there could be some argument to somehow test the jdk11 built artifacts on jdk8 but I guess that would take some extra work here.

The guidance on this stuff has been scattered and unclear in some ways. What I've outline here has been gathered from a wide variety of blog posts, stackoverflow comments, etc. If I'm misunderstanding something, please let me know.

**Comment by jorgebay** [comment]
> I think we should add a job below using openjdk8

I still think having a job with jdk 8 will give us the certainty that it actually runs on jdk 8. In my experience different runtimes sometimes exhibit quirks and different behaviors even if the API contract is the same.

**Comment by jorgebay** [comment]
It looks like last run went well!

VOTE +1

**Comment by spmallette** [comment]
Yes - it is working but doc generation is not. I'm glad I checked that. I'm not sure I had before.

**Comment by spmallette** [comment]
Docs are generating now. Have to re-run all the build options again to make sure everything is good. Hopefully no more problems.

**Comment by spmallette** [comment]
Docker builds are now unstable. Something strange with surefire/failsafe plugins. Seems like it is a known problem, but I'm not sure what is only triggering it now. This body of work never seems to find a bottom.

**Comment by spmallette** [comment]
I was able to get two good docker builds in a row with the latest commit. I had to get rid of my old docker images and rebuild incidentally. Continuing testing of other build options.

**Comment by spmallette** [comment]
Just got an error building locally now over surefire/failsafe. same issue as i was once having with docker:

```text
# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# A fatal error has been detected by the Java Runtime Environment:'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# A fatal error has been detected by the Java Runtime Environment:'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#  SIGSEGV (0xb) at pc=0x00007f74e7fb610b, pid=29287, tid=29309'.

# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# JRE version: OpenJDK Runtime Environment (11.0.6+10) (build 11.0.6+10)'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# JRE version: OpenJDK Runtime Environment (11.0.6+10) (build 11.0.6+10)'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# Java VM: OpenJDK 64-Bit Server VM (11.0.6+10, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)'.

# Created at 2020-06-27T06:40:14.303
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# Problematic frame:'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# Problematic frame:'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.310
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# V  [libjvm.so+0x85410b]  IndexSetIterator::advance_and_next()+0x11b'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# V  [libjvm.so+0x85410b]  IndexSetIterator::advance_and_next()+0x11b'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.311
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.311
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport %p %s %c %d %P %E" (or dumping to /home/smallette/git/apache/tinkerpop/neo4j-gremlin/core.29287)'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport %p %s %c %d %P %E" (or dumping to /home/smallette/git/apache/tinkerpop/neo4j-gremlin/core.29287)'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.311
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.311
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# An error report file with more information is saved as:'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# An error report file with more information is saved as:'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.311
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# /home/smallette/git/apache/tinkerpop/neo4j-gremlin/hs_err_pid29287.log'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# /home/smallette/git/apache/tinkerpop/neo4j-gremlin/hs_err_pid29287.log'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.333
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.333
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# Compiler replay data is saved as:'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# Compiler replay data is saved as:'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.333
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# /home/smallette/git/apache/tinkerpop/neo4j-gremlin/replay_pid29287.log'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '# /home/smallette/git/apache/tinkerpop/neo4j-gremlin/replay_pid29287.log'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.333
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.333
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '# If you would like to submit a bug report, please visit:'.

# Created at 2020-06-27T06:40:14.333
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#   https://github.com/AdoptOpenJDK/openjdk-support/issues'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#   https://github.com/AdoptOpenJDK/openjdk-support/issues'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)


# Created at 2020-06-27T06:40:14.333
Corrupted STDOUT by directly writing to native stream in forked JVM 1. Stream '#'.
java.lang.IllegalArgumentException: Stream stdin corrupted. Expected comma after third character in command '#'.
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient$OperationalData.<init>(ForkClient.java:507)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.processLine(ForkClient.java:210)
	at org.apache.maven.plugin.surefire.booterclient.output.ForkClient.consumeLine(ForkClient.java:177)
	at org.apache.maven.plugin.surefire.booterclient.output.ThreadedStreamConsumer$Pumper.run(ThreadedStreamConsumer.java:88)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by vtslab** [comment]
Sounds frustrating, indeed. The following fix was merged in failsafe-2.2.2: https://github.com/apache/maven-surefire/commit/18b4078de24a0cd69eb51cad4e793d4e51b26743 and I see and older version in the TinkerPop pom.xml. Worth a try.

**Comment by spmallette** [comment]
With some added changes, this seems to be working now. Running more tests today and then expect to immediately merge.

**Comment by LifeIsStrange** [comment]
Why not Java 13 or 14? Java 15 will reach general availability this september and the generated bytecode will still be compatible for clients having an openjdk 8.
Java devs has a culture of lagging years behing the stable releases despite having a language designed with excellent backward compatibility in mind.
I feel that java would loose less marketshare over C#, Go and others competitors if like them, the devs used the actual language and not a decade old version of it. 

Anyway sorry for my rant it has nothing specific to do with tinkerpop but I wanted to say it. 
This PR is still an improvement but feels to me a little bit too little too late.

**Comment by spmallette** [comment]
@LifeIsStrange we can only move at the pace allowed by downstream libraries we depend on. spark 3.0 just only released a few weeks ago with jdk11 support. hadoop is still working on proper jdk11 support https://cwiki.apache.org/confluence/display/HADOOP/Hadoop+Java+Versions but luckily doesn't affect our build, but does affect local documentation generation, so we could move forward without it. Never mind issues with maven plugin infrastructure and jdk11 support.

**Comment by LifeIsStrange** [comment]
@spmallette thanks for the anwser:
*we can only move at the pace allowed by downstream libraries we depend on.* This is the part I do not understand, I had the belief that you can set a javac version for your codebase and that each of your external dependencies can be compiled with their own version of javac. Because your more recent version is backward compatible it should not be an issue.
I might be wrong but it would be weird, other ecoystems like the rust programming language do not have the issue of requiring their compiler version to be the most recent version supported by all used dependencies.

**Comment by spmallette** [comment]
Maybe it is all as you stated and I'm sorry that I can't provide too much information to support/refute it given my experience...a bit too much time has passed since I started this a year ago for me to remember.

I'm just happy it's all working at this point and that this thing will finally merge. 😄 :shipit:

**Comment by porunov** [comment]
@LifeIsStrange there are small breaking changes in Java 9, 10 and 11. The things are a little bit more complicated now because depending on what your `jar` uses (which was compiled with jdk 8), it now may work or may not work in javac 9, 10 or 11. 
Thus, the project can't easily be updated to later java if it uses some old deps which won't work.

@spmallette Thank you for such a great update!

**Comment by darkduckgo** [comment]
@spmallette sorry to disturb, with the java 17 (the latest LTS version) published

it's important for graph database like neo4j/JanusGraph which is based on Java env to improve the `GC` perf with the stable `zgc` or `oneGC` maybe. (especially for low latency or large memory env)

So will tinkerpop have the plan to support build in java17? (the Java11 will end lifecycle in 2026), or provide a TODO list for adopt it, we can help to solve them together.

BTW, @porunov  will Janus consider support Java17 in next release version? or in a roadmap maybe

**Comment by spmallette** [comment]
personally, I have no problem with advancing the version of JDK support but if our dependencies don't support the JDK version we want to target then that can cause problems. i havent' looked at jdk17 yet to know what those problems would be. so far there is at least this issue to sort out: https://issues.apache.org/jira/browse/TINKERPOP-2633 further discussion on this topic should probably happen there or on the dev list.

**Comment by LifeIsStrange** [comment]
@spmallette I don't know enough about the topic but I was wondering wether JDK 9 modules eliminate conflicts between libraries java runtime versions?
https://en.m.wikipedia.org/wiki/Java_Platform_Module_System

**Comment by spmallette** [comment]
There is an open issue about JPMS somewhere - I've not had time to investigate it. I probably need to read up more on how that should/could be used. Of course, we're still building to to java 8 compatible bytecode even on jdk11. We really haven't taken advantage of any any non-jdk8 features as of yet.

**Pull request merged**